### PR TITLE
fix(helm): better file type detection

### PR DESCRIPTION
**Pull request by Zebradil**
## 📑 Description

This PR improves filetype detection for helm files.

The current filetype detection often enables yaml-ls for helm templates. yaml-ls then marks helm specific syntax as errors. To avoid this, a few specific rules are added to detect helm files.

## ℹ Additional Information

The new code is mainly based on the following sources:

- https://github.com/mandos/nvim-helm/blob/7b5b5945bf9c84b17dceb2f9a6cabfac7a6d12f1/lua/helm/init.lua
- https://github.com/mandos/nvim-helm/blob/7b5b5945bf9c84b17dceb2f9a6cabfac7a6d12f1/lua/helm/utils.lua

They are under the MIT license. How should I attribute them?

- [x] Add attribution to the source code
- [x] Test the changes (I have tested them but as a separate module, outside of astrocommunity)

### CODE REVIEW

File: lua/astrocommunity/pack/helm/init.lua
Comment (Uzaaft) [inline lua/astrocommunity/pack/helm/init.lua:16]: Why isn't this just moved into the astrocore spec?

File: lua/astrocommunity/pack/helm/init.lua
Comment (Uzaaft) [inline lua/astrocommunity/pack/helm/init.lua:33]: These can be moved into the opts, and convert the opts stuff to the function notation.

File: lua/astrocommunity/pack/helm/init.lua
Comment (Zebradil) [reply lua/astrocommunity/pack/helm/init.lua:16]: Sorry, I actually meant to do that.

File: lua/astrocommunity/pack/helm/init.lua
Comment (Zebradil) [reply lua/astrocommunity/pack/helm/init.lua:33]: Are there any examples on how to do that? My lua proficiency level is "copy someone else's code around".

Comment (Zebradil) [reply lua/astrocommunity/pack/helm/init.lua:33]: Ah, wait a second, I think I got it.

Comment (Zebradil) [reply lua/astrocommunity/pack/helm/init.lua:18]: Did you mean something like this?

```lua
  {
    "AstroNvim/astrocore",
    ---@type AstroCoreOpts
    opts = {
      autocmds = {
        helm_commentstring = {
          {
            event = "FileType",
            pattern = "helm",
            callback = function() vim.opt_local.commentstring = "{{/* %s */}}" end,
          },
        },
      },
      filetypes = {
        extension = {
          gotmpl = "helm",
          yaml = function(path, _) return M.is_helm_file(path) and "helm.yaml" or "yaml" end,
          yml = function(path, _) return M.is_helm_file(path) and "helm.yaml" or "yaml" end,
          tmpl = function(path, _) return M.is_helm_file(path) and "helm.tmpl" or "template" end,
          tpl = function(path, _) return M.is_helm_file(path) and "helm.tmpl" or "smarty" end,
        },
        filename = {
          ["Chart.yaml"] = "yaml",
          ["Chart.lock"] = "yaml",
        },
        pattern = {
          ["helmfile.*%.ya?ml"] = "helm",
        },
      },
    },
  },
```

I still need to keep the `is_helm_file` function somewhere. And the function for yaml and yml extensions is repeated.

File: lua/astrocommunity/pack/helm/init.lua
Comment (Zebradil) [reply lua/astrocommunity/pack/helm/init.lua]: Removed.

File: lua/astrocommunity/pack/helm/init.lua
Comment (Uzaaft) [reply lua/astrocommunity/pack/helm/init.lua:18]: No. More like:
```lua
opts = function(_, opts)
-- Do opts stuff here
end
```

And define stuff inside the function.

Comment (Zebradil) [reply lua/astrocommunity/pack/helm/init.lua]: I have updated the code. Now opts is a function which defines local helper functions inside. Did I get your idea right?

**Comment by Uzaaft** [comment]
I'm just gonna ask for a second set of eyes from @AstroNvim/astrocommunity-maintainers

**Review by mehalter — COMMENTED** [review]
One thing we may need to start thinking about (and may require fixing other packs) is how we will have these compound file type detection functions play nicely with each other. The filetype detection conflicts with other language packs like the audible pack.

Also is it true that all gotmpl files are helm files 🤔 that seems wrong

**Comment by Zebradil** [comment]
> have these compound file type detection functions play nicely with each other.

I don't think it's possible to do this in an automated way. In cases of conflicts, users will have to resolve them on a per-case basis. With time and reports from other users, we can figure out better rules to minimize conflicts.

> Also is it true that all gotmpl files are helm files 🤔 that seems wrong

I don't know, this was in the pack [from the start](https://github.com/AstroNvim/astrocommunity/commit/f8fbe8404559e8314b4096cffa20e20c1f08fbfe), added by @Sacro.

I don't work a lot with non-standard Helm templates. So far, I have seen the following files that are Helm templates:

- `NOTES.txt` (`helm.tmpl`)
- `**/*.yaml` in a chart directory (`helm.yaml`)
- `helmfile.yaml` (`helm.yaml`)
- `something.yaml.gotmpl` (`helm.yaml`)

With the current rules in this PR, I have false positives where `values.yaml` is detected as `helm.yaml` instead of `yaml`. So I was thinking whether the rules should be changed to be more precise, but I probably haven't seen all possible cases so far to make a good guess about the best set of rules.

**Comment by mehalter** [comment]
> I don't think it's possible to do this in an automated way. In cases of conflicts, users will have to resolve them on a per-case basis. With time and reports from other users, we can figure out better rules to minimize conflicts.

I can fix this, I will push a fix to this branch on Monday when I get to a computer again. I will also fix the rest of the filetype additions throughout AstroCommunity

**Comment by Sacro** [comment]
> I don't know, this was in the pack [from the start](https://github.com/AstroNvim/astrocommunity/commit/f8fbe8404559e8314b4096cffa20e20c1f08fbfe), added by @Sacro.

I've slept since then, I've looked through Discord and I can't remember where that came from or why I left it in.

**Comment by mehalter** [comment]
@Zebradil I just pushed a commit that adds filetype merging as a utility in AstroCommunity and uses it for both the ansible pack and the helm pack. I have tested them together and they play nicely together! Also I reverted the change that all of these should map to the `helm` filetype and not `helm.yaml`/`helm.tmpl`. This breaks the language server and I can't find any example of those filetypes being used anywhere. Let me know what you think of the changes!

**Comment by Zebradil** [comment]
> I just pushed a commit that adds filetype merging as a utility in AstroCommunity and uses it for both the ansible pack and the helm pack.

Awesome, that's really neat 👍 

> I reverted the change that all of these should map to the `helm` filetype and not `helm.yaml`/`helm.tmpl`. This breaks the language server and I can't find any example of those filetypes being used anywhere.

Yeah, sorry for that. I inspected the origins of this idea (to use `helm.yaml` instead of `helm`) more carefully and it seems like it was not finished and it requires more configuration (to parser and to language server) to work. I'm puzzled why it didn't break when I was testing this, but I'd leave it for now. Maybe I figure that out and file another PR.

It'd be nice to merge this if there are no more suggestions. The question about code attribution is not relevant anymore, as there is nothing left from the original code.

**Review by mehalter — APPROVED** [review]

**Pull request merged**
